bpo-34454: Clean up datetime.fromisoformat surrogate handling

pganssle opened the pull request:
This is a fixup PR for #8862, per @serhiy-storchaka's comments. I have addressed two PEP-7 violations, switched to using `_PyUnicode_Copy` in `_sanitize_fromisoformat_str`, and fixed an issue where the sanitized string (rather than the original string) was displayed as part of the error message.

Additionally, I noticed that the pure Python implementation uses the equivalent of `%U` instead of `%R` in its error printing, so I switched all the `fromisoformat` errors over to using `%U` for consistency.

[bpo-34454](https://www.bugs.python.org/issue34454)

<!-- issue-number: [bpo-34454](https://www.bugs.python.org/issue34454) -->
https://bugs.python.org/issue34454
<!-- /issue-number -->

pganssle committed:
PEP 7 fixes to from_isoformat

pganssle committed:
Use _PyUnicode_Copy in sanitize_isoformat_str

pganssle committed:
Use repr in fromisoformat error message

This reverses commit 67b74a98b2 per Serhiy Storchaka's suggestion:

     I suggested to use %R in the error message because including the raw
     string can be confusing in the case of empty string, or string
     containing trailing whitespaces, invisible or unprintable characters.

We agree that it is better to change both the C and pure Python versions
to use repr.

pganssle committed:
Retain non-sanitized dtstr for error printing

This does not create an extra string, it just holds on to a reference to
the original input string for purposes of creating the error message.

taleinat commented on Lib/test/datetimetester.py:
```diff
@@ -2676,6 +2676,14 @@ def test_fromisoformat_fails_datetime(self):
+        # the separator, the error message contains the original string
+        dtstr = "2018-01-03\ud80001:0113"
+
+        with self.assertRaisesRegex(ValueError, f".*{dtstr}"):
```
IMO just checking that `ValueError` is raised is enough.

pganssle replied:
The regex is actually the point of the test. Before these changes, the error message was accidentally including the *sanitized* string, so `2018-01-03\ud80001:0113` would give you something like `Invalid isoformat str: 2018-01-03T01:0113`.

As a result of this test, I also realized that the C and pure Python versions had slightly different error messages, so I corrected that (and this test enforces it).

If we don't care to enforce any particular conditions on the error message, we can drop the whole test, since the part about raising `ValueError` has pretty good test coverage elsewhere.

taleinat reviewed (approved):
LGTM, one small comment.

zhangyangyu reviewed (commented):
Is the `needs_decref` flag a must to accomplish the logic? Isn't `Py_XDECREF` enough?

pganssle committed:
Separate handling of Unicode and other errors

In the initial implementation, errors other than encoding errors would
both raise an error indicating an invalid format, which would not be
true for errors like MemoryError.

serhiy-storchaka commented on Lib/test/datetimetester.py:
```diff
@@ -2681,9 +2681,12 @@ def test_fromisoformat_fails_surrogate(self):
+        with self.assertRaises(ValueError) as cm:
             self.theclass.fromisoformat(dtstr)
 
+        msg = cm.exception.args[0]
```
Or just `str(cm.exception)`.

pganssle commented on Lib/test/datetimetester.py:
```diff
@@ -2676,6 +2676,17 @@ def test_fromisoformat_fails_datetime(self):
+        with self.assertRaises(ValueError) as cm:
+            self.theclass.fromisoformat(dtstr)
+
+        msg = cm.exception.args[0]
```
I was having trouble using a `assertRaisesRegex` here, I think because of the escape characters, but `in` works fine.

zhangyangyu replied:
You could use `assertRaisesRegex(ValueError, re.escape(f"{dtstr!r}"))`.

pganssle replied:
Ah, thank you! I thought there must be a way to do this but it seemed like a lot of work to figure it out. Thanks for the pointer. :smile:

pganssle committed:
Drop needs_decref from _sanitize_isoformat_str

Instead _sanitize_isoformat_str returns a new reference, even to the
original string.

izbyshev reviewed (approved):
LGTM!

taleinat commented on Lib/test/datetimetester.py:
```diff
@@ -2676,6 +2677,14 @@ def test_fromisoformat_fails_datetime(self):
+        # the separator, the error message contains the original string
+        dtstr = "2018-01-03\ud80001:0113"
+
+        with self.assertRaisesRegex(ValueError, re.escape(f"{dtstr!r}")):
```
Perahps just `re.escape(repr(dtstr))`?

pganssle replied:
Oops, thanks. This is a legacy from the fact that I wasn't sure if this had to match at the beginning of a line, so it was `f".*{dtstr!}"`.

pganssle replied:
Fixed. Thanks.

vstinner reviewed (changes requested):
Mixing a bugfix and code cleanup is a bad practice: it becomes harder to cherry-pick the fix to older versions, it's harder to revert the change in case of failure, etc.

Please revert all coding style changes and create a new PR once this one is merged.

I'm ok with a PR only fixing coding style, since I like all your coding styles changes, it really enhances the code :-)

vstinner reviewed (approved):
LGTM.
